### PR TITLE
Add sortWith to List / Add unfolds to List, Map and Set

### PR DESCRIPTION
**Pull request by stephentetley**
Pull request for issue #1052 - Add `sortWith` to List / Add unfolds to List, Map and Set

Adds the following functions:

~~~

List.flix - sortWith
List.flix - unfold
List.flix - unfoldWithIter

Map.flix - unfold
Map.flix - unfoldWithIter

Set.flix - unfold
Set.flix - unfoldWithIter

~~~

The `sortWith` is a Flix implementation of `sortBy` from Haskell's Data.List - the Haskell code has been refined at least three times for performance.

### CODE REVIEW

**Comment by stephentetley** [comment]
D'oh, the last commit message should have said "Changed unfoldWithIterator to unfoldWithIter".

File: main/src/library/List.flix [hunk 1]
```diff
@@ -844,6 +844,115 @@ namespace List {
             a
         }
 
+    ///
+    /// Sort list `xs` with the comparing function `cmp`.
+    ///
+    /// The sort implementation is a mergesort.
+    /// Acknowlegdement: derived from Thomas Nordin's `sortBy` in the Haskell base libraries.
```
Comment (magnus-madsen) [inline main/src/library/List.flix:851]: Acknowledgement

File: main/src/library/List.flix [hunk 2]
```diff
@@ -844,6 +844,115 @@ namespace List {
...
+    /// `f` should return `Some(a,st1)` to signal a new list element `a` and a new seed value `st1`.
+    ///
+    /// `f` should return `None` to signal the end of building the list.
+    ///
+    pub def unfold(f: s -> Option[(a, s)] & e, st: s): List[a] & e = unfoldHelper(f, st, xs -> xs)
+
+    ///
+    /// Helper function for `unfold`.
+    ///
+    def unfoldHelper(f: s -> Option[(a, s)] & e, st: s, cont: List[a] -> List[a]): List[a] & e = match f(st) {
```
Comment (magnus-madsen) [inline main/src/library/List.flix:933]: I prefer the name `k` for continuations.

File: main/src/library/List.flix [hunk 3]
```diff
@@ -844,6 +844,115 @@ namespace List {
...
+
+    ///
+    /// Build a list by applying the function `next` to `()`. `next` is expected to encapsulate
+    /// a stateful resource such as a file handle that can be iterated.
+    ///
+    /// `next` should return `Some(a)` to signal a new list element `a`.
+    ///
+    /// `next` should return `None` to signal the end of building the list.
+    ///
+    pub def unfoldWithIter(next: () -> Option[a] & e): List[a] & e = unfoldWithIterHelper(next, xs -> xs)
```
Comment (magnus-madsen) [inline main/src/library/List.flix:946]: Is this not a great example of a situation where you expect an impure function? :)

If `next` is pure then unfolding it with `unfoldWithIter` makes no sense, right? 

You can require an impure function by writing `next: () ~> Option[e]`.

The only other example I have found of a similar use case is `List.foreach` (and similar), so this is very cool!

File: main/src/library/Map.flix [hunk 4]
```diff
@@ -493,6 +493,41 @@ namespace Map {
...
+    /// `f` should return `Some(k,v,st1)` to signal a new key-value pair `k` and `v` and a new seed value `st1`.
+    ///
+    /// `f` should return `None` to signal the end of building the map.
+    ///
+    pub def unfold(f: s -> Option[(k, v, s)] & e, st: s): Map[k,v] & e = unfoldHelper(f, st, empty())
+
+    ///
+    /// Helper function for `unfold`.
+    ///
+    pub def unfoldHelper(f: s -> Option[(k, v, s)] & e, st: s, ac: Map[k,v]): Map[k,v] & e = match f(st) {
```
Comment (magnus-madsen) [inline main/src/library/Map.flix:508]: Non-public.

File: main/src/library/Map.flix [hunk 5]
```diff
@@ -493,6 +493,41 @@ namespace Map {
...
+
+    ///
+    /// Build a map by applying the function `next` to `()`. `next` is expected to encapsulate
+    /// a stateful resource such as a file handle that can be iterated.
+    ///
+    /// `next` should return `Some(k,v)` to signal a new key-value pair `k` and `v`.
+    ///
+    /// `next` should return `None` to signal the end of building the map.
+    ///
+    pub def unfoldWithIter(next: () -> Option[(k, v)] & e): Map[k, v] & e = unfoldWithIterHelper(next, empty())
```
Comment (magnus-madsen) [inline main/src/library/Map.flix:521]: Demand an impure function.

File: main/src/library/Set.flix [hunk 6]
```diff
@@ -334,6 +334,41 @@ namespace Set {
...
+    /// `f` should return `Some(a,st1)` to signal a new set element `a` and a new seed value `st1`.
+    ///
+    /// `f` should return `None` to signal the end of building the set.
+    ///
+    pub def unfold(f: s -> Option[(a, s)] & e, st: s): Set[a] & e = unfoldHelper(f, st, empty())
+
+    ///
+    /// Helper function for `unfold`.
+    ///
+    pub def unfoldHelper(f: s -> Option[(a, s)] & e, st: s, ac: Set[a]): Set[a] & e = match f(st) {
```
Comment (magnus-madsen) [inline main/src/library/Set.flix:349]: Non-public.

File: main/src/library/Set.flix [hunk 7]
```diff
@@ -334,6 +334,41 @@ namespace Set {
...
+
+    ///
+    /// Build a set by applying the function `next` to `()`. `next` is expected to encapsulate
+    /// a stateful resource such as a file handle that can be iterated.
+    ///
+    /// `next` should return `Some(a)` to signal a value pair `a`.
+    ///
+    /// `next` should return `None` to signal the end of building the set.
+    ///
+    pub def unfoldWithIter(next: () -> Option[a] & e): Set[a] & e = unfoldWithIterHelper(next, empty())
```
Comment (magnus-madsen) [inline main/src/library/Set.flix:362]: Demand an impure function.

**Comment by stephentetley** [comment]
Thanks Magnus, I've just pushed those changes but got a git error about fast-forwards. I think it seems to have gone through though.

**Comment by magnus-madsen** [comment]
Thanks Stephen!

**Pull request merged**